packages/services/src/kernel/serialize.ts

```diff
@@ -74,10 +74,14 @@ namespace Private {
     );
     let buffers = [];
     for (let i = 5; i < offsets.length; i++) {
+      let buffer: DataView;
       if (i == offsets.length - 1) {
-        buffers.push(new DataView(binMsg.slice(offsets[i])));
+        buffer = new DataView(binMsg.slice(offsets[i]));
       } else {
-        buffers.push(new DataView(binMsg.slice(offsets[i], offsets[i + 1])));
+        buffer = new DataView(binMsg.slice(offsets[i], offsets[i + 1]));
+      }
+      if (buffer.byteLength > 0) {
```

Comment:
Is this safe? Seems like you could in fact push empty arrays (for whatever reason) and those should be deserialized correctly. In Panel I tried a test case where I pushed an empty array and that (correctly) got deserialized as a zero-length DataView (which appeared in addition to the bogus additional `DataView`).

Reply:
Here's what I mean, this is a message where I've sent an empty array:

<img width="851" alt="Screen Shot 2023-06-20 at 14 32 49" src="https://github.com/jupyterlab/jupyterlab/assets/1550771/4d31c038-38eb-4a35-bf2a-1b776f098d11">

You can see that we get two `DataView` with length 0, one being the real empty array and then the bogus empty `DataView` that is always added.

Reply:
Yes I can see a need for inserting empty buffers. Could you try this PR with your Panel example?

Reply:
Having some trouble and about to be in meetings for a few hours. `jupyter lab build` fails without any indication of what went wrong.

Reply:
Maybe run `jlpm run build`?

Reply:
... and `jlpm install` to install the npm deps.

https://jupyterlab.readthedocs.io/en/latest/developer/contributing.html#setting-up-a-local-development-environment should contain the ordered list of commands to run, which is documented on https://jupyterlab.readthedocs.io/en/latest/developer/contributing.html#installing-jupyterlab

```
git clone https://github.com/<your-github-username>/jupyterlab.git
cd jupyterlab
pip install -e ".[dev,test]"
jlpm install
jlpm run build  # Build the dev mode assets (optional)
```

I haven't tried those simple commands in a fresh env (cc/ @ericsnekbytes)

Reply:
Actually, according to [this comment](https://github.com/jupyter-server/jupyter_server/pull/657#issuecomment-1021202444):
> `offsetN == len(binMsg)` (i.e., the last offset is the length of the file, so that the last buffer is `binMsg[offset[N-1], offset[N]]`?

I don't think we need to process the last offset at all. That's what I did in https://github.com/jupyterlab/jupyterlab/pull/14721/commits/ce3c98aa3d6ed5b119d7d25aa6e709c2b3e337b4.
Maybe @jasongrout can confirm?

Reply:
That was my feeling too, it seems like the simplest fix. In all the messages I've manually stepped through it also would have produced the correct result. For example in my screenshot above where I sent an empty array you can see that the accumulated offsets ends with `..., 547, 547]` i.e. it correctly records the offset for the empty array and then adds an extra one that should not be processed.

Reply:
> Maybe @jasongrout can confirm?

I think that is right (see also https://github.com/jupyter-server/jupyter_server/pull/657#issuecomment-1021406374). IIRC, the reason we made the last offset the length of the message is so that the decoding would be very simple, and it provides a sanity check on the message size, and lets us possibly include content after the last buffer in the future.